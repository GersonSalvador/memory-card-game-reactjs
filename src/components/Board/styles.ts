import styled from 'styled-components';

interface BoardProps {
  boardWidth: number;
}

export const Container = styled.div<BoardProps>`
  width: ${({ boardWidth }) => boardWidth + 'px'};
  display: flex;
  border: 1px solid var(--main-2);
  flex-wrap: wrap;
  justify-content: center;
  align-items: center;
  background-color: var(--main-3);
  padding: 10px 5px;
  transition: all 0.8s ease-in;
`;

// interface liProps {
//   selected: boolean;
// }

// export const Container = styled.li<liProps>`
  
//   background: ${({ selected }) => (selected && 'rgba(0,0,0,.2)')};