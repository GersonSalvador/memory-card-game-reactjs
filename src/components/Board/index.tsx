import { useEffect, useState} from "react";
import Card from "../Card";
import { Container } from "./styles"

interface BoardProps {
  levelInfo: {
    pairs: number;
    width: number;
  }
}

function Board({levelInfo: {pairs, width}}: BoardProps){
  const cardsIndexArr = [...Array(pairs)].map((item, index) => index)
  const pairsIndexArr = [...cardsIndexArr, ...cardsIndexArr]
  const [cardsIndex, setCardsIndex] = useState<Number[]>([])

  useEffect(() => {
    const shuffled = pairsIndexArr.sort(() => Math.random() - 0.5);
    setCardsIndex(shuffled)
  }, [])
  return (
    <Container data-testid="board" boardWidth={width}>
      {
        cardsIndex.map((i, idx) => <Card img={`https://avatars.dicebear.com/api/bottts/${i}.svg`} key={`img-${idx}`}/>)
      }
    </Container>
  )
}

export default Board;