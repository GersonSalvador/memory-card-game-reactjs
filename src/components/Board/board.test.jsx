import { render, act } from '@testing-library/react';
import renderer from 'react-test-renderer'
import 'jest-styled-components'
import Board from './';

describe('Board component', () => {
  afterEach(() => {
    jest.resetAllMocks();
  });
  const renderMyComponent = () => {
    const setIsFinished = jest.fn();
    const setIsWon = jest.fn();
    const setIsStarted = jest.fn();
    return {
      render: render(<Board 
        levelInfo={{pairs: 1, width: 450}}
        setIsFinished={setIsFinished}
        setIsWon={setIsWon}
        setIsStarted={setIsStarted}
      />),
      setIsFinished,
      setIsWon,
      setIsStarted,
    };
  }

  it('renders on screen', () => {
    const {render: {getByTestId}} = renderMyComponent();
    const component = getByTestId('board');
    expect(component).toBeInTheDocument();
  });

  it('renders 1 pair of cards', () => {
    const  {render: {getAllByTestId}} = renderMyComponent();
    const cards = getAllByTestId('card');
    expect(cards.length).toBe(2);
  });

  it('should render Board with 450px width', () => {
    const setIsFinished = jest.fn();
    const setIsWon = jest.fn();
    const tree = renderer.create(<Board 
      levelInfo={{pairs: 1, width: 450}}
      setIsFinished={setIsFinished}
      setIsWon={setIsWon}
    />).toJSON();

    expect(tree).toMatchSnapshot()
    expect(tree).toHaveStyleRule('width', '450px');
  });

  describe('Update States', () => {

    it('should update isFinished, isWon and isStaerted states when all cards are matched', async () => {
      const {
        render: {getAllByTestId}, 
        setIsFinished, 
        setIsWon,
        setIsStarted,
      } = renderMyComponent();
      const cards = getAllByTestId('card');
      await act(async () => {
        cards[0].click();
      });
      expect(setIsFinished).toHaveBeenCalledTimes(0)
      await act(async () => {
        cards[1].click();
        await new Promise((r) => setTimeout(r, 1100));
      });
      expect(setIsFinished).toBeCalledWith(true);
      expect(setIsFinished).toHaveBeenCalledTimes(1)
      expect(setIsWon).toBeCalledWith(true);
      expect(setIsWon).toHaveBeenCalledTimes(1);
      expect(setIsStarted).toHaveBeenCalledTimes(1);
      expect(setIsStarted).toBeCalledWith(false);
    });

  });
})