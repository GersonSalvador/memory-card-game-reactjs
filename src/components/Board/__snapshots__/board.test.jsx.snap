// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Board component should render Board with 450px width 1`] = `
.c0 {
  width: 450px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  border: 1px solid var(--main-2);
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: var(--main-3);
  padding: 10px 5px;
  -webkit-transition: all 0.8s ease-in;
  transition: all 0.8s ease-in;
}

<div
  className="c0"
  data-testid="board"
/>
`;
